crates/proof-of-sql/src/sql/postprocessing/group_by_postprocessing.rs

```diff
@@ -0,0 +1,325 @@
+use super::{PostprocessingError, PostprocessingResult};
+use indexmap::{IndexMap, IndexSet};
+use proof_of_sql_parser::{
+    intermediate_ast::{AggregationOperator, AliasedResultExpr, Expression},
+    Identifier,
+};
+use serde::{Deserialize, Serialize};
+
+/// A group by expression
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct GroupByPostprocessing {
+    /// A list of `AliasedResultExpr` that exclusively use identifiers in the group by clause or results of aggregation expressions
+    remainder_exprs: Vec<AliasedResultExpr>,
+
+    /// A list of identifiers in the group by clause
+    group_by_identifiers: Vec<Identifier>,
+
+    /// An `IndexMap` of aggregation expressions
+    aggregation_expr_map: IndexMap<(AggregationOperator, Expression), Identifier>,
+}
+
+/// Check whether multiple layers of aggregation exist within the same GROUP BY clause
+/// since this is not allowed in SQL
+///
+/// If the context is within an aggregation function, then any aggregation function is considered nested.
+/// Otherwise we need two layers of aggregation functions to be nested.
+fn contains_nested_aggregation(expr: &Expression, is_agg: bool) -> bool {
+    match expr {
+        Expression::Column(_) | Expression::Literal(_) | Expression::Wildcard => false,
+        Expression::Aggregation { expr, .. } => is_agg || contains_nested_aggregation(expr, true),
+        Expression::Binary { left, right, .. } => {
+            contains_nested_aggregation(left, is_agg) || contains_nested_aggregation(right, is_agg)
+        }
+        Expression::Unary { expr, .. } => contains_nested_aggregation(expr, is_agg),
+    }
+}
+
+/// Get identifiers NOT in aggregate functions
+fn get_free_identifiers_from_expr(expr: &Expression) -> IndexSet<Identifier> {
+    match expr {
+        Expression::Column(identifier) => IndexSet::from([*identifier]),
+        Expression::Literal(_) | Expression::Aggregation { .. } | Expression::Wildcard => {
+            IndexSet::new()
+        }
+        Expression::Binary { left, right, .. } => {
+            let left_identifiers = get_free_identifiers_from_expr(left);
+            let right_identifiers = get_free_identifiers_from_expr(right);
+            left_identifiers
+                .union(&right_identifiers)
+                .cloned()
+                .collect::<IndexSet<_>>()
```

Comment:
You can make this clone/copy free:
```suggestion
            let mut left_identifiers = get_free_identifiers_from_expr(left);
            let right_identifiers = get_free_identifiers_from_expr(right);
            left_identifiers.extend(right_identifiers);
            left_identifiers
```

Reply:
Alternatively, you could pass a `&mut IndexSet<Identifier>` into the method that builds this, which would avoid having to merge these altogether.